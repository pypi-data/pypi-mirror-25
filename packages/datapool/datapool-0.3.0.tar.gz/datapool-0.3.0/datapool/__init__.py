# -*- coding: utf-8 -*-

__author__ = 'Uwe Schmitt'
__email__ = 'uwe.schmitt@id.ethz.ch'
__version__ = '0.3.0'
