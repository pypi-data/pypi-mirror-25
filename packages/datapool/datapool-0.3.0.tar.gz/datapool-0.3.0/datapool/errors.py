# encoding: utf-8
from __future__ import print_function, division, absolute_import


class DataPoolException(Exception):
    pass


class ConsistencyError(DataPoolException):
    pass


class FormatError(DataPoolException):
    pass


class InvalidOperationError(DataPoolException):
    pass


class InvalidScriptName(DataPoolException):
    pass


class InvalidLandingZone(DataPoolException):
    pass


class InvalidConfiguration(DataPoolException):
    pass


class DataPoolIOError(DataPoolException):
    pass
