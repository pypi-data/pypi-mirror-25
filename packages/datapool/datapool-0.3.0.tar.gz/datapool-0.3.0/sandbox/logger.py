# encoding: utf-8
from __future__ import print_function, division, absolute_import

from datetime import datetime as dt
from functools import partial

import zmq


class Logger:
    """This is sending signals to the log_receiver via port 5559 (using zmq)
    """

    def __init__(self, context, id_=""):
        self.id_ = id_
        self.sender = context.socket(zmq.PUSH)
        self.sender.connect("tcp://127.0.0.1:5559")

        self.debug = partial(self._log, "debug")
        self.info = partial(self._log, "info")
        self.warn = partial(self._log, "warn")
        self.error = partial(self._log, "error")

    def shutdown(self):
        self.sender.send(b"KILL|")

    def _log(self, level, *a):
        text = " ".join(map(str, a))
        #text = "%-12s: %s" % (self.prefix, text)
        message = "%s|%s|%s|%s" % (level, dt.now(), self.id_, text)
        self.sender.send(message.encode("utf-8"), flags=zmq.NOBLOCK)
