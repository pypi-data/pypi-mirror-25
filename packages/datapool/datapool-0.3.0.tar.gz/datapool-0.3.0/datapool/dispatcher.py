# encoding: utf-8
from __future__ import print_function, division, absolute_import

from datetime import datetime
import functools
import os
import shutil
import time

from .database import connect_to_db
from .data_conversion import ConversionRunner, SUPPORTED_EXTENSIONS
from .errors import DataPoolException, InvalidOperationError
from .domain_objects import DomainObject
from .landing_zone_structure import (lookup_handler, script_paths_for_raw_file,
                                     source_yaml_path_for_script)

from .logger import logger
from .signal_model import commit as commit_signals, check_signals_against_db
from .uniform_file_format import to_signals
from .utils import iter_to_list
from .yaml_parsers import parse_source


def log_unexpected_exceptions(function):
    @functools.wraps(function)
    def wrapped_function(*a, **kw):
        logger()
        try:
            return function(*a, **kw)
        except Exception as e:
            args = list("{!r}".format(arg) for arg in a)
            args += list("{}={!r}".format(k, v) for (k, v) in sorted(kw.items()))
            call_signature = "{}({})".format(function.__name__, ", ".join(args))
            logger().error("unexpected error '{}' when calling {}".format(e, call_signature))

    return wrapped_function


def default_time_provider():
    return datetime.now().strftime("%Y%m%d-%H%M%S-%f")


class Dispatcher:

    def __init__(self, config, time_provider=default_time_provider):
        """
        time_provider arg allows patching for regression tests !
        """
        self.config = config
        self.time_provider = time_provider
        self.root_folder = config.landing_zone.folder
        self.conversion_runner = ConversionRunner(config)
        self.engine = connect_to_db(self.config.db)

    @iter_to_list
    def dispatch(self, rel_path):
        if rel_path.endswith(".yaml"):
            yield from self.dispatch_yaml(rel_path)
        elif rel_path.endswith(".raw"):
            yield from self.dispatch_raw_file_conversion(rel_path)
        elif any(rel_path.endswith(ext) for ext in SUPPORTED_EXTENSIONS):
            # no dispatch because dispatch is triggered on .raw file creation, only backup:
            yield from self.backup(rel_path)
        else:
            yield "do not know how to process file {}".format(rel_path)

    def backup(self, rel_path):
        try:
            yield from self._backup(rel_path)
        except IOError as e:
            yield "backup of {} failed: {}".format(rel_path, e)

    def _backup(self, rel_path):
        backup_path = os.path.join(self.config.backup_landing_zone.folder, rel_path)
        folder = os.path.dirname(backup_path)
        if not os.path.exists(folder):
            os.makedirs(folder)
            yield "created folder {}".format(folder)

        # add time stamp to make file unique and better searchable:
        backup_path += ".{}".format(self.time_provider())

        shutil.copy(self.full_path(rel_path), backup_path)
        yield "copied {} to {}".format(rel_path, backup_path)

    def full_path(self, *a):
        return os.path.join(self.root_folder, *a)

    def rel_path(self, p):
        return os.path.relpath(p, self.root_folder)

    def dispatch_raw_file_conversion(self, rel_raw_file_path):
        raw_file_path = self.full_path(rel_raw_file_path)
        script_paths = script_paths_for_raw_file(raw_file_path)

        if len(script_paths) > 1:
            yield InvalidOperationError("multiple conversion scripts found for {}"
                                        .format(rel_raw_file_path))
            return

        if len(script_paths) == 0:
            yield InvalidOperationError("no conversion script found next to {}"
                                        .format(rel_raw_file_path))
            return

        yield from self.run_conversion(script_paths[0], raw_file_path)
        yield from self.backup(rel_raw_file_path)

    def run_conversion(self, script_path, raw_file_path):
        rel_script_path = self.rel_path(script_path)
        rel_raw_file_path = self.rel_path(raw_file_path)

        msg = "convert {} with {}".format(rel_raw_file_path, rel_script_path)
        yield msg
        logger().info(msg)

        started = time.time()
        try:
            rows = self.conversion_runner.run_conversion(script_path, raw_file_path)
        except DataPoolException as e:
            msg = "converting {} with {} failed: '{}'".format(rel_raw_file_path, rel_script_path, e)
            yield msg
            logger().error(msg)
            return
        needed = time.time() - started

        msg = "timing: converting {} needed {:.2f} seconds".format(rel_raw_file_path, needed)
        yield msg
        logger().info(msg)

        source = next(parse_source(self.root_folder, source_yaml_path_for_script(script_path)))
        for row in rows:
            row["source"] = source.name

        started = time.time()
        signals = to_signals(rows)
        errors = check_signals_against_db(signals, self.engine)
        if errors:
            MAXE = 50
            too_many = len(errors) > MAXE
            orig_len = len(errors)
            errors = errors[:MAXE]
            for e in errors:
                msg = "converting {} with {} failed: '{}'".format(rel_raw_file_path,
                                                                  rel_script_path, e)
                yield InvalidOperationError(msg)
                logger().error(msg)
            if too_many:
                msg = "skipped {} error messages".format(orig_len - MAXE)
                yield InvalidOperationError(msg)

            return
        msg = "checked {} signals, needed {:.2f} seconds".format(len(signals), needed)
        yield msg
        logger().info(msg)

        started = time.time()
        commit_signals(signals, self.engine)
        needed = time.time() - started
        msg = "commited {} signals, needed {:.2f} seconds".format(len(signals), needed)
        logger().info(msg)
        yield msg

    def dispatch_yaml(self, rel_path):
        handler = lookup_handler(rel_path)

        if handler is None:
            yield InvalidOperationError("don't know how to handle {}".format(rel_path))
            return

        ok = True
        domain_object = None
        for result in handler.parser(self.root_folder, rel_path):
            if isinstance(result, DomainObject):
                domain_object = result
                continue
            yield InvalidOperationError("when parsing {}: {}".format(rel_path, result))
            ok = False

        if not ok:
            return

        if domain_object is None:
            yield "nothing to commit"
            return

        try:
            committed = handler.committer(domain_object, self.engine)
            if committed:
                yield "commited changes"
            else:
                yield "nothing committed, possibly no update"
        except DataPoolException as e:
            yield e

        self.backup(rel_path)
