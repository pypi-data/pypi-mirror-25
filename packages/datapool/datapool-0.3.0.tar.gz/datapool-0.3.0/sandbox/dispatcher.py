# encoding: utf-8

import os
import signal
import subprocess
import sys
import threading
import time

import zmq

from logger import Logger


# for debugging: print stacktrace if SIGUSR1 is received:
import faulthandler
faulthandler.register(signal.SIGUSR1)


def run_worker(*args):
    # it is better to use absolution imports here, but for development this will not work
    # unless we implement a command line to start the dispatcher
    try:
        from . import worker
    except:
        import worker
    worker_path = os.path.abspath(worker.__file__)
    cmdline = [sys.executable, worker_path] + list(args)

    def preexec():
        signal.signal(signal.SIGINT, signal.SIG_IGN)

    # here the argument preexec_fn avoids forwarding of KeyboardInterrupt to workers
    return subprocess.Popen(cmdline, bufsize=0, preexec_fn=preexec)


class Dispatcher:
    """ The dispatcher is receiving work from demo.py (aka the client) via 'inproc://client' (in-process communication)
    A poller is introduced, which watches both for incoming work and available workers (communicating via port 5555).
    If there is incoming work, a worker is popped from the list of available workers and the work is sent to that worker.
    When the worker is finished, it sends the message "READY" back to the poller; the worker is then added to the list of available workers again.

    
    At any time, log-info is sent to the logger which is sending the message to the log_receiver (port 5559)
    """

    def __init__(self, num_workers, context, worker_ip="127.0.0.1", worker_port=5555):
        self.num_workers = num_workers

        self.workers_connect_to = "tcp://%s:%d" % (worker_ip, worker_port)
        self.client_sends_to = "inproc://client"

        self.context = context
        self.logger = Logger(context, "dispatcher")

        self._start_workers()
        self._setup_client_queue()

        self.dispatch_thread = threading.Thread(target=self._dispatch_loop)
        # self.dispatch_thread.daemon = True  # else Ctrl-C will not work
        self.dispatch_thread.start()

    def _start_workers(self):
        procs = []
        for worker_index in range(self.num_workers):
            id_ = "worker_%d" % worker_index
            proc = run_worker(id_, self.workers_connect_to)
            procs.append((id_, proc))

        self.procs = procs

    def _setup_client_queue(self):
        self.queue_socket = self.context.socket(zmq.PUSH)
        self.queue_socket.connect(self.client_sends_to)

    def submit(self, data):
        self.queue_socket.send(b"PROCESS|%s" % data)

    def shutdown(self):
        self.queue_socket.send(b"DONE|")
        self.logger.info("wait for dispatch thread to join")
        #while self.dispatch_thread.isAlive():
        #    time.sleep(.001)

        # wait until all threads are finished
        self.dispatch_thread.join()
        self.queue_socket.close()

    def terminate_workers(self):
        for id_, p in self.procs:
            p.wait()
            self.logger.info("%s (pid=%d) finished" % (id_, p.pid))

    def _dispatch_loop(self):

        self.frontend = self.context.socket(zmq.PULL)
        self.frontend.bind(self.client_sends_to)

        self.backend = self.context.socket(zmq.ROUTER)
        self.backend.bind(self.workers_connect_to)

        poller = zmq.Poller()
        poller.register(self.backend, zmq.POLLIN)
        poller.register(self.frontend, zmq.POLLIN)

        available_workers = []
        self._wait_for_all_workers(available_workers, poller)
        self.logger.info("available workers: %d" % len(available_workers))

        self._dispatch_all_work(available_workers, poller)
        self._wait_until_all_workers_are_available(available_workers, poller)

        self._send_kill_to_all_workers(available_workers)
        self._terminate()

    def _wait_for_all_workers(self, available_workers, poller):
        missing = self.num_workers
        while missing:

            socks = dict(poller.poll())

            if socks.get(self.backend) == zmq.POLLIN:
                worker_id, empty, message = self.backend.recv_multipart()
                assert empty == b""
                assert message == b"READY"
                missing -= 1
                available_workers.append(worker_id)

        self.logger.info("all %d workers up" % len(available_workers))

    def _dispatch_all_work(self, available_workers, poller):

        while True:

            socks = dict(poller.poll())

            if socks.get(self.backend) == zmq.POLLIN:
                self._update_available_workers(available_workers)

            # we only check for data to process if we have at least one worker available:
            if socks.get(self.frontend) == zmq.POLLIN:
                if available_workers:
                    action, payload = self.frontend.recv().split(b"|", 1)
                    self.logger.info("dispatcher received: %s %r" % (action, payload))
                    if action == b"DONE":
                        break
                    else:
                        # send payload to be processed by one available worker:
                        worker_id = available_workers.pop()
                        msg = b"PROCESS|%s" % payload
                        self.backend.send_multipart([worker_id, b"", msg])
                else:
                    # got data but no worker available
                    time.sleep(.001)

    def _update_available_workers(self, available_workers):
        worker_id, empty, message = self.backend.recv_multipart()
        assert empty == b""
        assert message == b"READY"
        available_workers.append(worker_id)

    def _wait_until_all_workers_are_available(self, available_workers, poller):
        # we got KILL and wait until all workers are ready again:
        self.logger.info("num workers", self.num_workers)
        while len(available_workers) < self.num_workers:
            self.logger.info("avail workers", available_workers)
            socks = dict(poller.poll())
            if socks.get(self.backend) == zmq.POLLIN:
                self._update_available_workers(available_workers)
            else:
                time.sleep(.001)
        self.logger.info("all workers waiting")

    def _send_kill_to_all_workers(self, available_workers):

        while available_workers:
            worker_id = available_workers.pop()
            msg = b"KILL"
            self.backend.send_multipart([worker_id, b"", msg])
            self.logger.info("sent KILL to %s" % worker_id)

        self.logger.info("sent KILL to all workers")

    def _terminate(self):
        self.logger.info("terminate worker processes")
        self.terminate_workers()
        self.frontend.close()
        self.backend.close()
        self.logger.info("zmq frontend + backend closed")
