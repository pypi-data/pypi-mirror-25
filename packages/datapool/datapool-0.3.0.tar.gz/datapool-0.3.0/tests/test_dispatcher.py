# encoding: utf-8
from __future__ import print_function, division, absolute_import

from datapool.database import setup_db, dump_db
from datapool.dispatcher import Dispatcher, log_unexpected_exceptions
from datapool.utils import find_executable
from datapool.logger import logger


def test_log_unexpected_exceptions(regtest_logger):
    @log_unexpected_exceptions
    def by_zero(a, b, c):
        return a + b / c

    with regtest_logger():
        by_zero(1, 2, 0)
        by_zero("1", 2, c=0)
        by_zero({1}, b="2", c=0)


def _run_dispatcher(config, regtest_logger, lz, pathes):
    setup_db(config.db)
    with regtest_logger():
        dispatcher = Dispatcher(config, time_provider=lambda: "NOW")
        for path in pathes:
            for message in dispatcher.dispatch(path):
                logger().info("message: {}".format(message))


def test_dispatcher_ok(config, regtest, regtest_logger, lz):
    pathes = (
        "sites/example_site/site.yaml",
        "data/parameters.yaml",
        "data/FloDar/source_type.yaml",
        "data/FloDar/FloDar_python/source.yaml",
        "data/FloDar/FloDar_python/conversion.py",
        "data/FloDar/FloDar_python/raw_data/data-001.raw", )

    config.python.executable = find_executable("python")
    _run_dispatcher(config, regtest_logger, lz, pathes)
    dump_db(config.db, file=regtest, max_rows=25)


def test_dispatcher_no_python(config, regtest_logger, lz):
    # we don't configure r here to trigger an error
    pathes = ("data/FloDar/FloDar_python/conversion.py",
              "data/FloDar/FloDar_python/raw_data/data-001.raw")

    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_no_raw(config, regtest_logger, lz):
    lz.remove_file("data/FloDar/FloDar_python/raw_data/data-001.raw")
    pathes = ("data/FloDar/FloDar_python/conversion.py", )
    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_no_script(config, regtest_logger, lz):
    lz.remove_file("data/FloDar/FloDar_python/conversion.py")
    pathes = ("data/FloDar/FloDar_python/raw_data/data-001.raw", )
    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_multiple_scripts(config, regtest_logger, lz):
    lz.overwrite_file("data/FloDar/FloDar_python/conversion.jl",
                      lz.p("data/FloDar/FloDar_python/conversion.py"))
    pathes = ("data/FloDar/FloDar_python/raw_data/data-001.raw", )
    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_wrong_yaml_name(config, regtest_logger, lz):
    pathes = ("data/invalid_name.yaml", )
    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_invalid_yaml(config, regtest_logger, data_path, lz):
    lz.overwrite_file("data/parameters.yaml", data_path("yamls/empty.yaml"))
    pathes = ("data/parameters.yaml", )
    _run_dispatcher(config, regtest_logger, lz, pathes)
