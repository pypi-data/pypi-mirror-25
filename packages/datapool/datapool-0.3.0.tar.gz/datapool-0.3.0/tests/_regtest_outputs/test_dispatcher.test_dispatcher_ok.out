INFO     eawag.datapool  connected to db sqlite+pysqlite:///<tmpdir>/datapool.db
INFO     eawag.datapool  load <tmpdir>/landing_zone/sites/example_site/site.yaml
DEBUG    eawag.datapool  called check for site test_site
INFO     eawag.datapool  called commit for site test_site
DEBUG    eawag.datapool  add new site row <SiteDbo name=test_site, city=Schacht 24G3S, Mischsystem, x=580000, y=255300, z=455.3, description=All the details of the site.All the details the site.All the details of the site.All the details of the site.All the details of the site.All the details of the site.All the details of the site.>
INFO     eawag.datapool  message: commited changes
INFO     eawag.datapool  load <tmpdir>/landing_zone/data/parameters.yaml
INFO     eawag.datapool  number of parameters found: 14
INFO     eawag.datapool  called _commit for parameter Absorbance 200.00_nm
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Absorbance 200.00_nm, unit=m-1, description=absorbance at 200.00 nm wavelength>
INFO     eawag.datapool  called _commit for parameter Absorbance 202.50_nm
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Absorbance 202.50_nm, unit=m-1, description=absorbance at 202.50 nm wavelength>
INFO     eawag.datapool  called _commit for parameter Average Flow Velocity
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Average Flow Velocity, unit=m/s, description=None>
INFO     eawag.datapool  called _commit for parameter Battery Voltage
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Battery Voltage, unit=V, description=None>
INFO     eawag.datapool  called _commit for parameter Distance
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Distance, unit=km, description=None>
INFO     eawag.datapool  called _commit for parameter Distance Reading Count
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Distance Reading Count, unit=1, description=None>
INFO     eawag.datapool  called _commit for parameter Flow
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Flow, unit=m/s, description=None>
INFO     eawag.datapool  called _commit for parameter Number of Samples
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Number of Samples, unit=1, description=None>
INFO     eawag.datapool  called _commit for parameter Peak to Mean Ratio
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Peak to Mean Ratio, unit=1, description=None>
INFO     eawag.datapool  called _commit for parameter Surcharge Level
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Surcharge Level, unit=m, description=None>
INFO     eawag.datapool  called _commit for parameter Surface Flow Velocity
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Surface Flow Velocity, unit=m/s, description=None>
INFO     eawag.datapool  called _commit for parameter Temperature
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Temperature, unit=C, description=None>
INFO     eawag.datapool  called _commit for parameter Water Level
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=Water Level, unit=m, description=None>
INFO     eawag.datapool  called _commit for parameter rain intensity
DEBUG    eawag.datapool  add new parameter row <ParameterDbo name=rain intensity, unit=mm/h, description=None>
INFO     eawag.datapool  message: commited changes
INFO     eawag.datapool  load <tmpdir>/landing_zone/data/FloDar/source_type.yaml
DEBUG    eawag.datapool  called commit for source_type FloDar
INFO     eawag.datapool  called _commit for source_type FloDar
DEBUG    eawag.datapool  add new source_type row <SourceTypeDbo name=FloDar, description=description of FloDar>
INFO     eawag.datapool  message: commited changes
INFO     eawag.datapool  load <tmpdir>/landing_zone/data/FloDar/FloDar_python/source.yaml
INFO     eawag.datapool  load <tmpdir>/landing_zone/data/FloDar/source_type.yaml
DEBUG    eawag.datapool  called commit for source flodar_python
DEBUG    eawag.datapool  added new averaging:
DEBUG    eawag.datapool  ParameterAveraging:
DEBUG    eawag.datapool      parameter            : 'Temperature'
DEBUG    eawag.datapool      integration_length_x : None
DEBUG    eawag.datapool      integration_length_y : None
DEBUG    eawag.datapool      integration_angle    : None
DEBUG    eawag.datapool      integration_time     : 3600
DEBUG    eawag.datapool  added new averaging:
DEBUG    eawag.datapool  ParameterAveraging:
DEBUG    eawag.datapool      parameter            : 'Distance'
DEBUG    eawag.datapool      integration_length_x : 100
DEBUG    eawag.datapool      integration_length_y : None
DEBUG    eawag.datapool      integration_angle    : None
DEBUG    eawag.datapool      integration_time     : None
DEBUG    eawag.datapool  add new source_dbo <SourceDbo name=flodar_python, serial=1234567-980-3, description=source_description>
INFO     eawag.datapool  message: commited changes
INFO     eawag.datapool  message: created folder <tmpdir>/backup_landing_zone/data/FloDar/FloDar_python
INFO     eawag.datapool  message: copied data/FloDar/FloDar_python/conversion.py to <tmpdir>/backup_landing_zone/data/FloDar/FloDar_python/conversion.py.NOW
INFO     eawag.datapool  convert data/FloDar/FloDar_python/raw_data/data-001.raw with data/FloDar/FloDar_python/conversion.py
INFO     eawag.datapool  timing: converting data/FloDar/FloDar_python/raw_data/data-001.raw needed <TIME> seconds
INFO     eawag.datapool  load <tmpdir>/landing_zone/data/FloDar/FloDar_python/source.yaml
INFO     eawag.datapool  load <tmpdir>/landing_zone/data/FloDar/source_type.yaml
INFO     eawag.datapool  checked 14558 signals, needed <TIME> seconds
INFO     eawag.datapool  commited 14558 signals, needed <TIME> seconds
INFO     eawag.datapool  message: convert data/FloDar/FloDar_python/raw_data/data-001.raw with data/FloDar/FloDar_python/conversion.py
INFO     eawag.datapool  message: timing: converting data/FloDar/FloDar_python/raw_data/data-001.raw needed <TIME> seconds
INFO     eawag.datapool  message: checked 14558 signals, needed <TIME> seconds
INFO     eawag.datapool  message: commited 14558 signals, needed <TIME> seconds
INFO     eawag.datapool  message: created folder <tmpdir>/backup_landing_zone/data/FloDar/FloDar_python/raw_data
INFO     eawag.datapool  message: copied data/FloDar/FloDar_python/raw_data/data-001.raw to <tmpdir>/backup_landing_zone/data/FloDar/FloDar_python/raw_data/data-001.raw.NOW
table comment:

   comment_id     signal_id     text     timestamp     author
   ----------     ---------     ----     ---------     ------

table parameter:

   parameter_id     name                       description                            unit
   ------------     ----------------------     ----------------------------------     ----
   1                Absorbance 200.00_nm       absorbance at 200.00 nm wavelength     m-1
   2                Absorbance 202.50_nm       absorbance at 202.50 nm wavelength     m-1
   3                Average Flow Velocity      None                                   m/s
   4                Battery Voltage            None                                   V
   5                Distance                   None                                   km
   6                Distance Reading Count     None                                   1
   7                Flow                       None                                   m/s
   8                Number of Samples          None                                   1
   9                Peak to Mean Ratio         None                                   1
   10               Surcharge Level            None                                   m
   11               Surface Flow Velocity      None                                   m/s
   12               Temperature                None                                   C
   13               Water Level                None                                   m
   14               rain intensity             None                                   mm/h

table parameter_averaging:

   parameter_averaging_id     parameter_id     source_id     integration_length_x     integration_length_y     integration_angle     integration_time
   ----------------------     ------------     ---------     --------------------     --------------------     -----------------     ----------------
   1                          12               1             None                     None                     None                  3600.0
   2                          5                1             100.0                    None                     None                  None

table picture:

   picture_id     site_id     filename     description     timestamp               data
   ----------     -------     --------     -----------     -------------------     --------------------------------
   1              1           1.jpg        rainevent       2016-01-01 00:00:00     fab67a552c8d509f899fa6fe710852cf
   2              1           2.JPG        outside         2016-02-01 12:00:00     fab67a552c8d509f899fa6fe710852cf
   3              1           3.JPG        manhole         None                    fab67a552c8d509f899fa6fe710852cf

table project:

   project_id     title     description
   ----------     -----     -----------

table quality:

   quality_id     flag     method
   ----------     ----     ------

table signal:

   signal_id     value      timestamp               parameter_id     source_id     site_id     coord_x     coord_y     coord_z
   ---------     ------     -------------------     ------------     ---------     -------     -------     -------     -------
   1             159.25     2013-11-13 10:06:00     13               1             None        682558      239404
   2             157.5      2013-11-13 10:08:00     13               1             None        682558      239404
   3             155.75     2013-11-13 10:10:00     13               1             None        682558      239404
   4             160.24     2013-11-13 10:12:00     13               1             None        682558      239404
   5             164.91     2013-11-13 10:14:00     13               1             None        682558      239404
   6             165.98     2013-11-13 10:16:00     13               1             None        682558      239404
   7             164.91     2013-11-13 10:18:00     13               1             None        682558      239404
   8             161.74     2013-11-13 10:20:00     13               1             None        682558      239404
   9             156.25     2013-11-13 10:22:00     13               1             None        682558      239404
   10            151.02     2013-11-13 10:24:00     13               1             None        682558      239404
   11            149.77     2013-11-13 10:26:00     13               1             None        682558      239404
   12            145.6      2013-11-13 10:28:00     13               1             None        682558      239404
   skipped 14533 rows ...
   14547         13.17      2013-11-14 23:00:00     4                1             None        682558      239404
   14548         13.16      2013-11-15 00:00:00     4                1             None        682558      239404
   14549         13.15      2013-11-15 01:00:00     4                1             None        682558      239404
   14550         13.16      2013-11-15 02:00:00     4                1             None        682558      239404
   14551         13.11      2013-11-15 03:00:00     4                1             None        682558      239404
   14552         13.15      2013-11-15 04:00:00     4                1             None        682558      239404
   14553         13.12      2013-11-15 05:00:00     4                1             None        682558      239404
   14554         13.15      2013-11-15 06:00:00     4                1             None        682558      239404
   14555         13.12      2013-11-15 07:00:00     4                1             None        682558      239404
   14556         13.14      2013-11-15 08:00:00     4                1             None        682558      239404
   14557         13.14      2013-11-15 09:00:00     4                1             None        682558      239404
   14558         13.14      2013-11-15 10:00:00     4                1             None        682558      239404

table signal_quality:

   signal_quality_id     signal_id     quality_id     timestamp     author
   -----------------     ---------     ----------     ---------     ------

table site:

   site_id     name          description                                  street                 postcode     city                           coord_x     coord_y     coord_z
   -------     ---------     ----------------------------------------     ------------------     --------     --------------------------     -------     -------     -------
   1           test_site     All the details of the site.All the deta     Flensburger Straße     None         Schacht 24G3S, Mischsystem     580000      255300      455.3
                             ils the site.All the details of the site
                             .All the details of the site.All the det
                             ails of the site.All the details of the
                             site.All the details of the site.

table source:

   source_id     source_type_id     name              description            serial            manufacturer          manufacturing_date
   ---------     --------------     -------------     ------------------     -------------     -----------------     ------------------
   1             1                  flodar_python     source_description     1234567-980-3     manufacturer_name     2016-11-22

table source_type:

   source_type_id     name       description
   --------------     ------     ---------------------
   1                  FloDar     description of FloDar

table special_value_definition:

   special_value_definition_id     source_type_id     description      categorical_value     numerical_value
   ---------------------------     --------------     ------------     -----------------     ---------------
   1                               1                  not a number     NA                    -1.0
   2                               1                                   INF                   -2.0

